src/NativeScripts.purs

```diff
@@ -21,3 +41,66 @@ instance Show NativeScriptHash where
 nativeScriptHash :: NativeScript -> Maybe NativeScriptHash
 nativeScriptHash ns = wrap <<< Hashing.nativeScriptHash <$> convertNativeScript
   ns
+
+-- | `SetChoice` is an internal type representing internal state of
+-- | `getMaximumSigners` algorithm. There are a number of values (key hashes)
+-- | from which we must choose exactly `choice`. It's like a suspended `NOfK`
+-- | choice. We track the alternatives alongside their count to adjust for
+-- | possible duplicates in `allChoice`.
+type SetChoice a = Array { set :: Set a, choice :: Int }
```

Comment:
Isn't `choice` always equal to `Set.size set`?

Reply:
Because the only instance where you create a `{ set :: Set a, choice :: Int }` other then the empty and singleton variants is in `allChoice`, where you take care to maintain this invariant.

Reply:
Initially I had another algorithm in mind, but now I realize that the choice is really the same as `size`. I will update the comment. But I'll keep the `choice` around, because it is `O(1)`.

Reply:
You only observe `choice` (other than for creating new `choice`) right at the end. You could just calculate the sizes at the end in `getMaximumSigners` (before `maximumBy`, if you care about the cost)

Reply:
Something like:

```haskell
type SetChoice a = Array (Set a)
...
choices >>> map (\set -> { set, size: size set }) >>> maximumBy (compare `on` _.size) >>> map _.size >>> fromMaybe 0
```

Reply:
I agree.